File: src/RTM.Api/Domain/Transaction.cs
using System;
using System.Text.Json.Serialization;

namespace RTM.Api.Domain;

/// <summary>
/// Status lifecycle of a transaction as defined by the JSON data model.
/// </summary>
public enum TransactionStatus
{
    Pending,
    Completed,
    Failed
}

/// <summary>
/// Core transaction entity — mirrors the JSON data model exactly (5 fields):
/// transactionId, amount, currency, status, timestamp (UTC ISO-8601).
/// No merchantId and no extra fields; any addition must be intentional.
/// Value object with a validating constructor to keep instances always valid.
/// </summary>
public sealed class Transaction
{
    public Guid TransactionId { get; }
    public decimal Amount { get; }
    public string Currency { get; }
    public TransactionStatus Status { get; }
    public DateTimeOffset Timestamp { get; }

    /// <summary>
    /// Validating constructor, also used by the JSON serializer when cache
    /// entries are deserialized (parameters drive the mapping by name).
    /// </summary>
    [JsonConstructor]
    public Transaction(
        Guid transactionId,
        decimal amount,
        string currency,
        TransactionStatus status,
        DateTimeOffset timestamp)
    {
        if (!IsValidTransactionId(transactionId))
            throw new ArgumentException("transactionId must be a non-empty GUID.", nameof(transactionId));

        if (amount < 0)
            throw new ArgumentException("amount must be >= 0.", nameof(amount));

        if (string.IsNullOrWhiteSpace(currency) || currency.Length != 3)
            throw new ArgumentException("currency must be a 3-letter ISO code (e.g. 'USD').", nameof(currency));

        if (timestamp == default)
            throw new ArgumentException("timestamp must be a valid date/time.", nameof(timestamp));

        TransactionId = transactionId;
        Amount = amount;
        Currency = currency;
        Status = status;
        Timestamp = timestamp;
    }

    private static bool IsValidTransactionId(Guid id) =>
        id != Guid.Empty;
}

File: client/src/types/transaction.ts
/**
 * Transaction domain model — mirrors the backend 5-field JSON contract
 * (see CLAUDE.md: no extra fields; status is the same enum).
 *
 * The backend serializes the status as its name (JsonStringEnumConverter),
 * so the TS union uses the exact PascalCase strings the API produces.
 */

export type TransactionStatus = 'Pending' | 'Completed' | 'Failed';

export interface Transaction {
  transactionId: string; // GUID string (lowercase, e.g. "aaaaaaaa-aaaa-...")
  amount: number;        // decimal, in the given currency
  currency: string;      // ISO 4217 3-letter code (e.g. "USD", "ILS")
  status: TransactionStatus;
  timestamp: string;     // UTC ISO-8601, e.g. "2026-08-21T12:00:00+00:00"
}

/** Raw payload posted to the ingestion API (field names match the API contract). */
export interface IngestTransactionRequest {
  transactionId: string;
  amount: number;
  currency: string;
  status: TransactionStatus;
  timestamp: string;
}

/** Narrowing helpers for the status enum in the UI. */
export const isFinalStatus = (status: TransactionStatus): boolean =>
  status === 'Completed' || status === 'Failed';
